File: beautifier/beauty.py
import re
import sys

sys.setrecursionlimit(1000*1000)


class Parser:
    def __rshift__(self, that):
        return p(self, that) ^ (lambda a: a[1])

    def __lshift__(self, that):
        return p(self, that) ^ (lambda a: a[0])

    def __xor__(self, that):
        return Action(self, that)


class nreg(Parser):
    def __init__(self, param):
        self.param = re.compile(param)

    def __call__(self, i):
        m = self.param.search(i)
        return None if m is None else [m.group(0), i[len(m.group(0)):]]


skip = nreg(r'^(\s|\(\*.*\*\))*')


class Or(Parser):
    def __init__(self, *params):
        self.params = map(lambda v: st(v) if isinstance(v, basestring) else v, params)

    def __call__(self, i):
        for v in self.params:
            r = v(i)
            if r is not None:
                return r
        return None


class nstr(Parser):
    def __init__(self, param):
        self.param = param

    def __call__(self, i):
        return None if not i.startswith(self.param) else [self.param, i[len(self.param):]]


def st(s):
    return p(skip, nstr(s))


class p(Parser):
    def __init__(self, *params):
        self.params = map(lambda v: st(v) if isinstance(v, basestring) else v, params)

    def __call__(self, i):
        rs = []
        for v in self.params:
            r = v(i)
            if r is None:
                return None
            rs.append(r[0])
            i = r[1]
        return [rs, i]


class Action(Parser):
    def __init__(self, thiz, action):
        self.thiz = thiz
        self.action = action

    def __call__(self, i):
        r = self.thiz(i)
        if r is None:
            return None
        else:
            r2 = self.action(r[0])
            return None if r2 is None else [r2, r[1]]


class opt(Parser):
    def __init__(self, thiz):
        self.thiz = st(thiz) if isinstance(thiz, basestring) else thiz

    def __call__(self, i):
        r = self.thiz(i)
        return [[], i] if r is None else r


class rep(Parser):
    def __init__(self, thiz):
        self.thiz = st(thiz) if isinstance(thiz, basestring) else thiz

    def __call__(self, i):
        rs = []
        while(True):
            r = self.thiz(i)
            if r is None:
                return [rs, i]
            rs.append(r[0])
            i = r[1]


def rep1(thiz):
    return rep(thiz) ^ (lambda p: None if len(p) < 1 else p)


class notp(Parser):
    def __init__(self, thiz):
        self.thiz = st(thiz) if isinstance(thiz, basestring) else thiz

    def __call__(self, i):
        return [[], i] if self.thiz(i) is None else None


def reg(s):
    return p(skip, nreg(s))


class Nest:
    def __init__(self, name, n):
        self.name = name
        self.n = n

    def __repr__(self):
        return self.name
NestP = Nest("NestP", 1)
NestM = Nest("NestM", -1)


def n(*a):
    return p(*a) ^ (lambda a: [NestP, a, NestM])


reg2 = re.compile(r'(^.*\n.*$)')
whiteSpace = re.compile(r'(^\s*$)')
reg1 = re.compile(r'\n')


def flat(a, rc):
    if isinstance(a, list):
        for i in a:
            rc = flat(i, rc)
    else:
        rc.append(a)
    return rc


def cnv(e):
    e = flat(e, [])
    e3 = []
    i = 0
    while(i < len(e)):
        s = [e[i]]
        if isinstance(s[0], basestring) and reg2.search(s[0]) is not None:
            s = []
            while(i < len(e)):
                s2 = e[i]
                if s2 is NestM:
                    e3.append(s2)
                elif whiteSpace.search(s2) is not None:
                    s.append(s2)
                else:
                    s.append(s2)
                    break
                i += 1
        i += 1
        e3.extend(s)

    nest = 0
    e4 = []
    for s in e3:
        if isinstance(s, Nest):
            nest += s.n
        else:
            m = reg2.search(s)
            if m is not None:
                s = reg1.sub("\n"+("  " * nest), m.group(0))
            e4.append(s)
    return "".join(e4)

keywords = reg(r"^(let|in|if|else|then|rec|begin|end|match|with|type)\b")

id = Or(
    notp(keywords) >> reg(r"^[_a-zA-Z0-9]+"),
    reg(r'^[+\-*\/.<>:@=][+\-*\/.<>:@=]*'),
    reg(r'^[,!]'),
    reg(r'^("(\\.|[^"])*")')
)

exp = p(lambda i: p(exps, rep(p(notp(";;") >> p(";"), exps)))(i))

exp1 = Or(
    p("begin", n(exp), "end"),
    p("(", n(opt(exp)), ")"),
    p("{", n(opt(exp)), "}"),
    p("[", n(opt(exp)), "]"),
    p("if", n(exp), "then", n(exp), "else", exp),
    p("let", n(opt("rec"), exp), "in", exp),
    p(Or("match", "try"), n(exp), "with", opt("|"), n(exp), rep(p("|", n(exp)))),
    p("function", opt("|"), n(exp), rep(p("|", n(exp)))),
    p("type", n(id, "=", opt("|"), n(exp), rep(p("|", n(exp)))), opt(";;")),
    p("type", n(id, "=", exp), opt(";;")),
    p("open", n(id, rep(p(".", id))), opt(";;")),
    id
)

exps = rep1(exp1)

regparse = re.compile(r"^[\s]+", re.M)


def parse(s):
    s = regparse.sub("", s)
    return cnv(exp(s))

File: beautifier/beauty.js
exports.parse = function(i){
  function startsWith(i, param) {
    var ilen = i.length;
    var len = param.length;
    if (len > ilen) return false;
    else return i.substring(0, len) == param;
  }

  function any_char(i) {
    if(i.length > 0) return [i.substring(0,1), i.substring(1)];
    else return null;
  }
  var str = function(s) { return p(skip, nstr(s));};


  Function.prototype.or=function(that) {
    var thiz = this;
    return function(i) {
      var e = thiz(i); if(e === null) return that(i);
      return e;
    };
  };

  function p() {
    var args = arguments;
    for(var i = 0; i < args.length;i++) {
      if(typeof(args[i])=="string") args[i]=str(args[i]);
    }
    return function(s) {
      var rs = [];
      for(var i = 0; i < args.length; i++) {
        var r = args[i](s);
        if(r === null) return null;
        rs.push(r[0]);
        s = r[1];
      }
      return [rs,s];
    };
  }

  Function.prototype.next=function(that) {
    var thiz = this;
    return function(i) {
      var r = p(thiz,that)(i); if(r === null) return r;
      return [r[0][1], r[1]];
    };
  };

  Function.prototype.prev=function(that) {
    var thiz = this;
    return function(i) {
      var r = p(thiz,that)(i); if(r === null) return r;
      return [r[0][0], r[1]];
    };
  };

  Function.prototype.action=function(f) {
    var thiz = this;
    return function(i) {
      var r = thiz(i); if(r === null) return r;
      return [f(r[0]), r[1]];
    };
  };

  function opt (thiz) {

    if(typeof(thiz)=="string")thiz=str(thiz);
    return function(i) {
      var r = thiz(i); if(r === null) return [null, i];
      return r;
    };
  }

  function rep (thiz) {
    return function (i) {
      var rs = [];
      while(true){
        var r = thiz(i); if (r === null) return [rs, i];
        rs.push(r[0]);
        i = r[1];
      }
    };
  }

  function rep1 (thiz) {
    return rep(thiz).action(function(p){
      if(p.length < 1) return null;
      return p;
    });
  }

  function notp (thiz) {
    return function(i) {
      var r = thiz(i); if(r === null) return ["", i];
      return null;
    };
  }

  function nstr (param) {
    return function(i) {
      //console.log("nstr "+param+" "+i)
      if(!startsWith (i, param)) return null;
      //console.log("nstr ok");
      return [param, i.substring(param.length)];
    };
  }

  function range(c1, c2) {
    return function(i) {
      if(i.length <= 0) return null;
      var c = i.charCodeAt(0); if (c1.charCodeAt(0) <= c && c <= c2.charCodeAt(0))
        return [i.substring(0, 1), i.substring(1)];
      return null;
    };
  }

  function nreg(param) {
    return function(i) {
      var m = i.match(param);
      if (m === null) return null;
      return [m[0],i.substring(m[0].length)];
    };
  }

  var skip = nreg(/^(\s|\(\*[\s\S]*\*\))*/);

  var reg = function(s) { return p(skip, nreg(s));};

  function NestP(){this.p=true;}
  function NestM(){this.m=true;}


  function n() {
    return p.apply(this,arguments).action(function(a){
      return [new NestP(),a,new NestM()];
    });
  }

  function flat(a) {
    if (a === null) return [];

    if(a instanceof Array) {
      var rc = [];
      for(var i in a) {
        rc = rc.concat(flat(a[i]));
      }
      return rc;
    }
    return [a];
  }

  function cnv(e){
    //console.log(JSON.stringify(e));
    var reg2 = /(^.*\n.*$)/;
    var whiteSpace = /(^\s*$)/;
    var nest = 0;
    var e2 = [];
    for(var i in e) {
      var s = e[i];
      if(typeof(s)=="string"){
        var m = s.match(/(^.*\n)(\s+$)/); if(m) s = m[1];
      }
      e2.push(s);
    }
    //console.log(JSON.stringify(e2));

    var e3 = [];
    for(i = 0; i < e2.length; i++) {
      var s = e2[i];
      if(typeof(s)=="string"){
        var m = s.match(reg2); if(m) {
          e = [];
          for(;i < e2.length; i++) {
            var s2 = e2[i];
            if(s2 instanceof NestM) {
              e3.push(s2);
              continue;
            }
            if(s2.match(whiteSpace)) {e.push(s2); continue; }
            // whitespace以外だ。
            e.push(s2);break;
          }
          //console.log("k "+JSON.stringify(e));

          e3 = e3.concat(e);
          continue;
        }
      }
      e3.push(s);
    }
    //console.log("kore "+JSON.stringify(e3));

    function makeString(s,n) {
      var rs = [];
      for(var i = 0; i < n; i++) {
        rs.push(s);
      }
      return rs.join("");
    }

    e2 = [];
    for (var i in e3) {
      var s = e3[i];

      if(s instanceof NestP) {nest++; continue; }
      if(s instanceof NestM) {nest--; continue; }
      var m = s.match(reg2); if(m) {
          //console.log("add nest "+nest);
          e2.push(m[0].replace(/\n/g, "\n"+makeString("  ",nest) ) );
          continue;
      }
      e2.push(s);
    }
    //console.log(JSON.stringify(e2));
    return e2.join("");
  }

  var keywords  = reg(/^(let|in|if|else|then|rec|begin|end|match|with)\b/);


  var id        = (notp(keywords).next(reg(/^[_a-zA-Z0-9]+/)))
                .or(reg(/^[+\-*\/.<>=:@]+/))
                .or(reg(/^[,!]/))
                .or(reg(/^("(\\.|[^"])*")/));
  var exp1      = p("begin", n(exp), "end")
                .or(p(p("match").or(p("try")), n(exp), "with", opt("|"), n(exp), rep(p("|", n(exp)))))
                .or(p("(", n(opt(exp)), ")"))
                .or(p("{", n(opt(exp)), "}"))
                .or(p("[", n(opt(exp)), "]"))
                .or(p("let", n(p(opt("rec"), exp)), "in", exp))
                .or(p("type", n(p(id, "=", exp)), ";;", exp))
                .or(p("type", n(id, "=", opt("|"), n(exp), rep(p("|", n(exp)))), ";;"))
                .or(p("if", n(exp), "then", n(exp), "else", exp))
                .or(id);
  var exps      = rep1(exp1);
  function exp(i) {
    return p(exps, rep(p(";", exps)))(i);
  }
  var e = exp(i);
    //console.log(JSON.stringify(e));
  return cnv(flat(e));
};
